[#1] 사용자 매칭 및 조회, 채팅방 생성

khoon9 opened the pull request:
# 구현 API
- 사용자
    - get 
        - 본인 제외 무작위 사용자 조회
        - 특정 사용자 조회
- 채팅방
    - post
        -  채팅방 생성

# TDD
- 사용자 매칭 시나리오에 따라 TDD 방식으로 테스트 및 구현 포함

# 추가
- NotFoundException 예외 추가 및 handler 설정
- BadRequestException 예외 추가 및 handler 설정
- validation starter dependency 추가

khoon9 committed:
test: 사용자 매칭 시나리오

시나리오
- 사용자가 매칭을 시작한다.
- 본인을 제외한 무작위 사용자를 반환한다.

테스트 대상
- 본인 이외의 무작위 사용자 정상 반환 여부 테스트.
- 매칭 대상이 발견되지 않을 경우 예외 발생 여부 테스트.

khoon9 committed:
feat: 사용자 매칭 시나리오

- 앞서 작성한 테스트를 만족시키는 메서드 구현

User Entity
- Table name "users"로 변경
- Profile과 1대1 연관관계 지정

Profile Entity
- 매칭 상대방이 주목할 사용자 특성 정보를 담은 테이블

에외 추가
- NotFoundException 예외 추가 및 handler 설정

khoon9 committed:
chore: validation starter dependency 추가

khoon9 committed:
chore: user entity 변경사항 반영

- 사용자 가입 profile 설정 부분 추가

khoon9 committed:
chore: 리팩토링 반영

- test 코드에 UserService mock 추가

khoon9 committed:
feat: 사용자 API 추가

- 본인 제외 무작위 사용자 호출 get API 추가
- 특정 사용자 호출 get API 추가

khoon9 committed:
feat: entity 추가

ChatRoom
- 사용자 매칭 후 방 생성의 결과로 생성될 채팅방

ChatRoomUser
- User와 ChatRoom 사이의 N:N 연관관계를 위한 중간 테이블

khoon9 committed:
feat: 예외 추가 및 handler 설정

- BadRequestException 예외 추가 및 handler 설정

khoon9 committed:
feat: 채팅방 생성 기능 추가

- 채팅방 생성 시에 2명 이상 존재하도록 설정. 만약 준수하지 않는다면 BadRequestException 발생.
- 채팅방 생성시 기재한 사용자들 중 존재하지 않는 사용자가 포함되어 있을 경우 NotFoundException 발생.

khoon9 committed:
feat 채팅방 API 추가

- 채팅방 생성 post API 추가

f-lab-green commented on src/test/java/flab/transtalk/user/service/user/UserMatchingServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+        User user = User.builder()
+                .id(2L)
+                .name("사용자2")
+                .birthDate(LocalDate.of(1999,01,01))
```
사소한 코멘트:
01은 사람이 읽기는 쉽지 않은 것 같아요.

khoon9 replied:
넵! 반영했습니다.

f-lab-green commented on src/test/java/flab/transtalk/user/service/user/UserMatchingServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+        // when & then
+        assertThatThrownBy(()->userMatchingService.getUserExcept(currentUserId))
+                .isInstanceOf(NotFoundException.class)
+                .hasMessageContaining("발견된 사용자가 없습니다.");
```
메세지를 상수화 시켜서 관리하면 좋을 것 같아요. 변경점이 하나가 되도록이요.

khoon9 replied:
넵! 변경점이 하나가 되도록 별도의 클래스를 정의하여 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/repository/UserRepository.java:
```diff
@@ -2,6 +2,14 @@
+import java.util.List;
 
 public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("SELECT u FROM User u WHERE u.id <> :userId ORDER BY function('RAND')")
```
데이터가 많아지면 성능 문제가 발생할 수 있을 것 같아요. 어떻게 하면 좋을지 고민해보세요.

khoon9 replied:
해당 부분은 추후 사용자 정보를 기반으로 추천 로직을 구성할 예정입니다. 이후 로직이 구체화되는 단계에서 성능 이슈도 함께 고려하여 반영하겠습니다. 감사합니다!

f-lab-green commented on src/main/java/flab/transtalk/user/domain/User.java:
```diff
@@ -24,4 +29,18 @@ public class User {
+    @Builder.Default
+    private List<ChatRoomUser> chatRoomUsers = new ArrayList<>();
+
+    public void setProfile(Profile profile) {
```
좀 더 방어적으로 작성하는 것도 좋아보여요.

```java
public void setProfile(Profile profile) {
    if (this.profile != null) {
        this.profile.setUser(null);
    }
    this.profile = profile;
    if (profile != null && profile.getUser() != this) {
        profile.setUser(this);
    }
}
```

khoon9 replied:
넵! 반영했습니다. 반영 과정에서 User entity와 setProfile 메소드 뿐만 아니라 이와 연관된 Profile entity의 setUser 메소드 또한 방어적으로 구성하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/dto/res/UserResponseDto.java:
```diff
@@ -19,12 +19,20 @@ public class UserResponseDto implements Serializable {
                 .id(dto.getId())
                 .name(dto.getName())
                 .birthDate(dto.getBirthDate())
+                .profile(
```
profile이 존재한다고 가정하고 작성하면 NPE와 친해지게 됩니다.

khoon9 replied:
넵! user의 profile 필드가 존재할 경우 기존과 같이 반환하되, 만약 profile 이 null일 경우 해당 필드를 null로 반환하도록 하여 반영하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/service/user/UserMatchingService.java:
```diff
@@ -0,0 +1,23 @@
+    public UserResponseDto getUserExcept(Long currentUserId) {
+        User user = userRepository.findRandomUserExcept(currentUserId);
+        if (user == null){
+            throw new NotFoundException("발견된 사용자가 없습니다.");
```
메세지 상수화 필요해보여요.

khoon9 replied:
넵! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/service/user/UserService.java:
```diff
@@ -0,0 +1,36 @@
+    public UserResponseDto getUser(Long userId) {
+        Optional<User> user = userRepository.findById(userId);
+        if (user.isEmpty()){
+            throw new NotFoundException("존재하지 않는 사용자입니다.");
```
메세지 상수화 필요해보여요.

khoon9 replied:
넵! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/service/user/UserSignUpService.java:
```diff
@@ -13,11 +15,18 @@ public class UserSignUpService {
+        User user = reqDto.toEntity();
+
+        Profile profile = Profile.builder()
+                .language(LanguageSelection.KOR)
```
하드코딩보단 Dto에서 넘어오는걸 받는건 어떨까요?

khoon9 replied:
넵! 서비스 레이어의 하드코딩 부분을 dto의 기본값 설정으로 대체하여 반영하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/controller/UserController.java:
```diff
@@ -14,10 +14,21 @@
+    }
+
+    // 주어진 사용자 제외, 무작위 사용자 호출
+    @GetMapping("/matching/{currentUserId}")
```
path에 명시하기보다는 인증은 횡단 관심사이므로 주입받는 형태로 구성해보는 것도 좋을 것 같아요.

khoon9 replied:
넵! 현재는 PathVariable을 통해 사용자 ID를 넘기고 있으나, 추후 로그인 인증 기능이 붙으면 현재 사용자 정보를 주입받아 처리하도록 하겠습니다. 감사합니다.

f-lab-green commented on src/main/java/flab/transtalk/translation/domain/ChatRoomUser.java:
```diff
@@ -0,0 +1,29 @@
+    )
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.EAGER)
```
EAGER가 꼭 필요한 상황일까요?

khoon9 replied:
아닙니다! LAZY로 수정하였습니다!

f-lab-green commented on src/main/java/flab/transtalk/translation/domain/ChatRoom.java:
```diff
@@ -0,0 +1,30 @@
+    @Column(name = "id", nullable = false)
+    private Long id;
+
+    @Column(name = "title", length = 50)
```
채팅방 제목이 필수 요소라면 nullable을 false로 명시해보는 것도 좋을 것 같네요.

khoon9 replied:
넵! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/common/exception/BadRequestExceptionHandler.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+
+@RestControllerAdvice
+public class BadRequestExceptionHandler {
```
처리할 것 마다 클래스가 생성되고 있는 것 같은데요. 전역 핸들러를 만들어보는 것은 어떨까요?

khoon9 replied:
넵! 반영했습니다. 반영 과정에서 사용자 정의 예외에 대한 핸들러와 기존에 원래부터 존재하던 예외에 대한 핸들러로 분리하여 전역 핸들러를 구성하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/translation/dto/res/ChatRoomResponseDto.java:
```diff
@@ -0,0 +1,36 @@
+@Value
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class ChatRoomResponseDto implements Serializable {
```
Serializable은 여기에 어떤 것 때문에 필요하신건가요.

khoon9 replied:
템플릿 코드로 인해 주입되었습니다. 현재 dto 중 Serializable이 필요없는 단순 REST API 응답 DTO들에 대하여 Serializable 명시를 취소하여 반영하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/translation/service/chatroom/ChatRoomCreateService.java:
```diff
@@ -0,0 +1,51 @@
+        if (users.size() != reqDto.getParticipantUserIds().size()){
+            throw new NotFoundException("존재하지 않는 사용자가 포함되어 있습니다.");
+        }
+        ChatRoom chatRoom = ChatRoom.builder()
+                .title(reqDto.getTitle())
+                .build();
+
+        List<ChatRoomUser> links = users.stream()
+                .map(user->ChatRoomUser.builder()
+                        .user(user)
+                        .chatRoom(chatRoom)
+                        .build()
+        ).collect(Collectors.toList());
+        chatRoom.getChatRoomUsers().addAll(links);
```
양방향 연관관계 명시를 위해 다음과 같은 수정은 어떨까요?

ChatRoom엔티티에 추가:
```java
public void addUser(User user) {
    ChatRoomUser chatRoomUser = ChatRoomUser.builder()
            .chatRoom(this)
            .user(user)
            .build();
    chatRoomUsers.add(chatRoomUser);
}```

서비스 로직:
```java
ChatRoom chatRoom = ChatRoom.builder()
        .title(reqDto.getTitle())
        .build();

users.forEach(chatRoom::addUser);
```

khoon9 replied:
넵! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/translation/service/chatroom/ChatRoomService.java:
```diff
@@ -0,0 +1,18 @@
+
+@Service
+@RequiredArgsConstructor
+public class ChatRoomService {
```
단순히 위임만 하고 있는데 직접 사용하는건 어떨까요.

khoon9 replied:
넵! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/translation/service/chatroom/ChatRoomCreateService.java:
```diff
@@ -0,0 +1,51 @@
+        }
+        List<User> users = userRepository.findAllByIdIn(reqDto.getParticipantUserIds());
+        if (users.size() != reqDto.getParticipantUserIds().size()){
+            throw new NotFoundException("존재하지 않는 사용자가 포함되어 있습니다.");
```
어떤 사용자가 존재하지 않는지 정보를 명시해야 디버깅에 용이할 것 같네요.

khoon9 replied:
넵! 입력받은 참가자가 존재하지 않는 경우와 중복된 경우가 모두 존재할 수 있다는 가정하에, NotFoundException 예외 클래스가 inputIds와 outputIds 필드를 가지도록 수정하여 반영하였습니다ㅏ.

khoon9 committed:
chore: 테스트 코드 수정

- 가독성을 위해 01을 1로 수정

khoon9 committed:
refactor: 예외 메시지 상수화

- 예외 메시지를 상수로 관리하여 변경 지점을 단일화함

khoon9 committed:
refactor: 예외 핸들러 분리

- 사용자 정의 예외 핸들러와 Validation 예외 핸들러로 분리 진행

khoon9 committed:
chore: 채팅방 entity 필드 설정 수정

- title 필드 nullable = false로 설정

khoon9 commented on src/test/java/flab/transtalk/user/service/user/UserMatchingServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+    void  testGetUserExcept_1(){
+        // given
+        Long currentUserId = 1L;
+        User user = User.builder()
```
넵! 테스트 대상이 user임을 명시할 수 있도록 제공해주신 코드대로 반영했습니다.

khoon9 committed:
chore: test 코드 수정

- user 객체에 대한 테스트임을 명시하기 위해 코드 수정 진행

khoon9 committed:
chore: User/profile entity 메소드 수정

- User entity의 setProfile 메소드와 Profile entity의 setUser 메소드 방어적으로 구성

khoon9 committed:
refactor: 서비스 레이어 재구성

- 단순 위임 형태의 서비스 코드 구조를, 해당 메소드를 직접 사용하는 구조로 개선함

khoon9 committed:
chore: 양방향 연관관계 명시

- ChatRoom 엔티티에 양방향 연관관계 명시를 위한 addUser 메소드 정의

khoon9 committed:
chore: 예외 클래스 필드 수정

- NotFoundException 클래스가 입력된 id와 찾은 id 인스턴스 변수를 가지도록 수정

khoon9 committed:
chore: 예외 클래스 필드 수정

- UserResponseDto의 user -> dto 변환 메소드 수정.
- 매개변수로 받은 user의 profile 필드가 비어있는지 여부에 따라 변환 동작이 다르도록 구현.

khoon9 committed:
chore: Serializable 여부 수정, fetch 옵션 수정

- dto 중 현재 Serializable이 필요없는 대상들에 대해 Serializable 명시 취소
- ChatRoomUser의 fetch 옵션 수정

khoon9 committed:
chore: 서비스 코드 하드코딩 제거

- 서비스 레이어의 하드코딩 부분을 dto의 기본값 설정으로 대체

f-lab-green reviewed (approved):
🚀
